graphql_compiler/macros/__init__.py

```diff
@@ -121,6 +157,10 @@ def perform_macro_expansion(macro_registry, graphql_with_macro, graphql_args):
         the returned values are guaranteed to be identical to the input query and args.
     """
     query_ast = safe_parse_graphql(graphql_with_macro)
+    schema_with_macros = get_schema_with_macros(macro_registry)
+    validation_errors = validate_schema_and_ast(schema_with_macros, query_ast)
+    if validation_errors:
+        raise GraphQLValidationError(u'String does not validate: {}'.format(validation_errors))
```

Comment:
This pattern is used multiple time in the code, (if validation_errors ...). I think it might be best to put this in validate_schema_and_ast. Up to you though.

Reply:
The tricky thing is that if `validate_schema_and_ast` raises the error, it will be unable to include the GraphQL  string that was passed in -- since it's analyzing an AST rather than the string itself. I think there are few enough places where this pattern is used that maybe it's okay that we have to raise the error outside the function that detects it.

I do have one nitpick observation to make, though. Imagine if this error bubbled up from some UI: `String does not validate`. Would you think it was a bug in the UI, or in the thing you wrote? I'd say it's probably a 50-50. How about we make it `The provided GraphQL input is not valid:` instead?

Reply:
It's only used twice. And we might want to do something else with the errors, other than raising. For example, you can write tests that not only assert that something is invalid, but that it's invalid for the right reason. Or maybe return all errors to the user, instead of just the first one.